### PR TITLE
Update SSO Enterprise Admin Form to Add API Key Management section for Identity Providers

### PR DESCRIPTION
**Pull request by mjriley**
## Product Description
Despite having front-end changes, this is all hidden behind a feature flag that no users should have enabled, and thus should be invisible.

## Technical Summary
Associated ticket: https://dimagi.atlassian.net/browse/SAAS-15360. This code aims to introduce the new fields for multi-view API keys, but does not commit those fields to the backend yet

## Feature Flag
All functionality will be hidden  behind the new `MULTI_VIEW_API_KEYS` toggle

### Safety story
I've locally tested this to ensure that the fields display when the toggle is on, but nothing changes when the toggle is off. I've verified that you can continue to update the identity provider when the toggle is off, and that none of the new fields are saved when the toggle is on.

### Automated test coverage

The changes to the form are validated in `TestSsoSamlEnterpriseSettingsForm`

### QA Plan

No QA at this point.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Review by biyeun — CHANGES_REQUESTED** [review]
please see comment about the checkbox field. will followup with updates to styleguide for the best approach

File: corehq/apps/sso/forms.py [hunk 1]
```diff
@@ -595,11 +599,99 @@ class BaseSsoEnterpriseSettingsForm(forms.Form):
...
     def get_primary_fields(self):
+        if self.allow_multi_view_api_keys:
+            fieldset = crispy.Fieldset(
+                _('API Key Management'),
+                checkbox_field('always_show_user_api_keys'),
+                checkbox_field('enforce_user_api_key_expiration', data_bind='checked: enforceExpiration'),
+                crispy.Div(
+                    crispy.Field(
+                        'max_days_until_user_api_key_expiration', data_bind="value: expirationLengthValue"),
+                    crispy.HTML(
```
Comment (biyeun) [inline corehq/apps/sso/forms.py:672]: offline on this one too, but as a reminder: I think the best approach here is to use `render_to_string()` with a template loading `hq_shared_tags`. This can then reference the standard form offsets with `{% css_action_class %}` so that we are not hardcoding the offsets here—especially something to avoid in a python file always.

making a note to myself include examples of this in our styleguide

File: corehq/apps/sso/forms.py [hunk 2]
```diff
@@ -642,6 +735,19 @@ def clean_entity_id(self):
...
+        if not enforce_expiration:
+            return None
+
+        return self.cleaned_data['max_days_until_user_api_key_expiration']
+
+    def clean(self):
+        if not self.has_error('max_days_until_user_api_key_expiration'):
+            enforces_expiration = bool(self.cleaned_data['enforce_user_api_key_expiration'])
+            if enforces_expiration and not self.cleaned_data.get('max_days_until_user_api_key_expiration', None):
+                self.add_error('max_days_until_user_api_key_expiration', gettext('a value must be specified'))
```
Comment (biyeun) [inline corehq/apps/sso/forms.py:749]: how does this error appear to the user? As I mentioned in Jing's PR, we historically have thrown validation errors in the `clean_<field_slug>` method due to the superior positioning of the error next to the field with an issue. It's more clear to the user when this is the case. I believe when we use the `clean` method, the errors appear as a bulleted list at the top of the form, which is a less ideal experience.

File: corehq/toggles/__init__.py [hunk 3]
```diff
@@ -1120,6 +1120,14 @@ def _ensure_valid_randomness(randomness):
     namespaces=[NAMESPACE_DOMAIN],
 )
 
+MULTI_VIEW_API_KEYS = StaticToggle(
```
Comment (biyeun) [inline corehq/toggles/__init__.py:1123]: somewhat nit, but generally it's preferred to PR the new toggles separately, as this file is often touched frequently and is a source of conflicts when other folks are introducing new toggles. Is it possible for you to cherry-pick this commit into a separate PR while we take time to address the feedback on this PR? Thanks!

File: corehq/apps/sso/views/enterprise_admin.py [hunk 4]
```diff
@@ -64,12 +66,16 @@ def page_context(self):
...
             'edit_idp_form': self.edit_enterprise_idp_form,
             'idp_slug': self.idp_slug,
+            'show_api_fields': self.show_api_fields(),
             'toggle_client_secret': (
                 self.identity_provider.protocol == IdentityProviderProtocol.OIDC
                 and self.identity_provider.client_secret
             ),
         }
 
+    def show_api_fields(self):
```
Comment (biyeun) [inline corehq/apps/sso/views/enterprise_admin.py:76]: generally for these toggle checks, `@property` is preferred

File: corehq/apps/sso/views/enterprise_admin.py [hunk 5]
```diff
@@ -109,7 +115,9 @@ def edit_enterprise_idp_form(self):
             return form_class(
                 self.identity_provider, self.request.POST, self.request.FILES
             )
-        return form_class(self.identity_provider)
+
+        api_fields_enabled = self.show_api_fields()
+        return form_class(self.identity_provider, allow_multi_view_api_keys=api_fields_enabled)
```
Comment (biyeun) [inline corehq/apps/sso/views/enterprise_admin.py:120]: I think this value needs to be passed into `form_class` above.

Also, curious as to the logic behind all the different names here? `api_fields_enabled`, `show_api_fields`, `allow_multi_view_api_keys` and why `show_api_fields`. Why are all these things named separately? And why was `show_api_fields()` not an `@property`?

Also, `show_api_fields` seems vague for this boolean. I feel that it could get confused with the api secrets / tokens we need for accessing the user-management APIs for each Identity Provider. I have a preference for `show_mutli_view_api_keys_fields`, but I also can see an argument for that being a bit too wordy.

File: corehq/apps/sso/forms.py (hunk 2 above)
Comment (biyeun) [reply corehq/apps/sso/forms.py:749]: rescinding my comment about `clean` as I just saw your comment about `self.add_error`.

Comment (biyeun) [reply corehq/apps/sso/forms.py:749]:
```suggestion
                self.add_error('max_days_until_user_api_key_expiration', gettext('Please specify a value.'))
```

File: corehq/apps/sso/forms.py [hunk 6]
```diff
@@ -553,6 +553,10 @@ def update_identity_provider(self, admin_user):
         return self.idp
 
 
+def checkbox_field(field_name, **kwargs):
```
[the remark this answers is not in the document]
Comment (biyeun) [reply corehq/apps/sso/forms.py:556]: see https://github.com/dimagi/commcare-hq/pull/34339

File: corehq/toggles/__init__.py (hunk 3 above)
Comment (mjriley) [reply corehq/toggles/__init__.py:1123]: Created a separate PR: https://github.com/dimagi/commcare-hq/pull/34352

File: corehq/apps/sso/views/enterprise_admin.py (hunk 4 above)
Comment (mjriley) [reply corehq/apps/sso/views/enterprise_admin.py:76]: I prefer to avoid properties if I can. Their primary purpose is to provide syntactic sugar when an object needs to conform to a specific interface. The main reason I believe they should be avoided is that they mask the complexity behind the call. When you write `object.property`, you expect no performance impact, as looking up a field on an object is trivial. However, `enabled_for_request` can ultimately result in calls to external systems (`toggle_enabled` can hit Couch, for example). This violates the law of least surprise.

File: corehq/apps/sso/views/enterprise_admin.py (hunk 5 above)
Comment (mjriley) [reply corehq/apps/sso/views/enterprise_admin.py]: Thanks for pointing out the inconsistencies in variable naming here. I've rearranged the code to avoid the property entirely and renamed the variables to all be `uses_api_key_management`, as that is the new section that is being introduced to the form for all of these properties. I realized I didn't like `show_*` because show is often read as a verb, which implies a function.

Both of the issues should be addressed in 8b43ece8ae4a67e73af1861bf8d9b53bc0982263

File: corehq/apps/sso/forms.py (hunk 2 above)
Comment (mjriley) [reply corehq/apps/sso/forms.py]: Addressed in 8b43ece8ae4a67e73af1861bf8d9b53bc0982263

File: corehq/apps/sso/forms.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mjriley) [reply corehq/apps/sso/forms.py]: Fixed in 2373f1ce58b5984e8166a8edf489e91a592f950a

File: corehq/apps/sso/forms.py (hunk 1 above)
Comment (mjriley) [reply corehq/apps/sso/forms.py]: Addressed in d36fdbacf03a01c74ccf7da2260370026db7fef9

File: corehq/apps/sso/views/enterprise_admin.py [hunk 7]
```diff
@@ -16,6 +16,8 @@
 )
 from corehq.apps.sso.models import IdentityProvider, IdentityProviderProtocol
 
+from corehq.toggles import MULTI_VIEW_API_KEYS
```
Comment (biyeun) [inline corehq/apps/sso/views/enterprise_admin.py:19]: generally the practice has been to use `from corehq import toggles` (which should already be in `views/enterprise_admin.py`...did you merge latest master into this branch yet) and then reference `toggles.MULTI_VIEW_API_KEYS`

File: corehq/toggles/__init__.py (hunk 3 above)
Comment (biyeun) [reply corehq/toggles/__init__.py:1123]: thanks!

File: corehq/apps/sso/forms.py [hunk 8]
```diff
@@ -616,13 +618,70 @@ class BaseSsoEnterpriseSettingsForm(forms.Form):
...
+            ('', '-----'),
+            (365, gettext_lazy("1 Year")),
+            (180, gettext_lazy("180 Days")),
+            (160, gettext_lazy("160 Days")),
+            (120, gettext_lazy("120 Days")),
+            (90, gettext_lazy("90 Days")),
+            (60, gettext_lazy("60 Days")),
+            (30, gettext_lazy("30 Days")),
+        ),
+        coerce=int,
```
Comment (biyeun) [inline corehq/apps/sso/forms.py:666]: nice!

**Comment by biyeun** [comment]
@mjriley I just ran this branch locally and it looks like for OIDC forms there are some strange formatting issues. Can you look into this? Thanks! Fyi, I ran a check on both `SAML` and `OIDC` without the feature flag.
Screen Shot 2024-04-02 at 6 52 40 PM

**Comment by biyeun** [comment]
@mjriley I just ran the branch locally and noticed some strange formatting issues with OIDC forms without the feature flag enabled.

**Comment by mjriley** [comment]
@biyeun the display issues should be addressed. It looks like I changed the base form to remove the redundant `<form>` tags and use the form helper to establish the form's CSS class -- and then didn't realize I needed to also update the OIDC's form helper as well. See d5609d1c052a3bd7190e18cbe4cdc73b07e7710c

**Review by biyeun — APPROVED** [review]
looks great! just that one comment about the form name for `OIDC` Identity Providers. thank you!

**Review by jingcheng16 — APPROVED** [review]
Looks good to me!
